File: public/.julia/v0.5/mosek/deps/src/mosek/8/tools/examples/fusion/java/callback.java
package com.mosek.fusion.examples;
/* 
   Copyright : Copyright (c) MOSEK ApS, Denmark. All rights reserved.

   File :      callback.java

   Purpose :   To demonstrate how to use the progress 
               callback. 

               Use this script as follows:
               
               callback psim  
               callback dsim  
               callback intpnt

               The first argument tells which optimizer to use
               i.e. psim is primal simplex, dsim is dual simplex
               and intpnt is interior-point. 
*/

import mosek.*;
import mosek.fusion.*;
import java.util.Formatter;

public class callback
{
  /*TAG:begin-callback*/                
  private static DataCallback makeUserCallback(final double maxtime,
                                               final Model mod)
    {
      return new DataCallback() {
          public int callback(callbackcode caller,
                              double[]     douinf,
                              int[]        intinf,
                              long[]       lintinf)
          {

            double opttime = 0.0;
            int itrn;
            double pobj, dobj, stime;

            Formatter f = new Formatter(System.out);
            /*TAG:begin-callback-switch*/                
            switch (caller)
            {
              case begin_intpnt:
                  f.format("Starting interior-point optimizer\n");
              case intpnt:
                  itrn    = intinf[iinfitem.intpnt_iter.value      ];
                  pobj    = douinf[dinfitem.intpnt_primal_obj.value];
                  dobj    = douinf[dinfitem.intpnt_dual_obj.value  ];
                  stime   = douinf[dinfitem.intpnt_time.value      ];
                  opttime = douinf[dinfitem.optimizer_time.value   ];
            /*TAG:end-callback-switch*/                
                  f.format("Iterations: %-3d\n",itrn);
                  f.format("  Time: %6.2f(%.2f) ",opttime,stime);
                  f.format("  Primal obj.: %-18.6e  Dual obj.: %-18.6e\n",pobj,dobj);
              case end_intpnt:
                  f.format("Interior-point optimizer finished.\n");
              case begin_primal_simplex:
                  f.format("Primal simplex optimizer started.\n");
              case update_primal_simplex:
                  itrn    = intinf[iinfitem.sim_primal_iter.value  ];
                  pobj    = douinf[dinfitem.sim_obj.value          ];
                  stime   = douinf[dinfitem.sim_time.value         ];
                  opttime = douinf[dinfitem.optimizer_time.value   ];
            
                  f.format("Iterations: %-3d\n", itrn);
                  f.format("  Elapsed time: %6.2f(%.2f\n",opttime,stime);
                  f.format("  Obj.: %-18.6e", pobj );
              case end_primal_simplex:
                  f.format("Primal simplex optimizer finished.\n");
              case begin_dual_simplex:
                  f.format("Dual simplex optimizer started.\n");
              case update_dual_simplex:
                  itrn    = intinf[iinfitem.sim_dual_iter.value    ];
                  pobj    = douinf[dinfitem.sim_obj.value          ];
                  stime   = douinf[dinfitem.sim_time.value         ];
                  opttime = douinf[dinfitem.optimizer_time.value   ];
                  f.format("Iterations: %-3d\n", itrn);
                  f.format("  Elapsed time: %6.2f(%.2f)\n",opttime,stime);
                  f.format("  Obj.: %-18.6e\n", pobj);
              case end_dual_simplex:
                  f.format("Dual simplex optimizer finished.\n");
              case begin_bi:
                  f.format("Basis identification started.\n");
              case end_bi:
                  f.format("Basis identification finished.\n");
              default:
            }
            System.out.flush();
            if (opttime >= maxtime)
            {
              f.format("MOSEK is spending too much time. Terminate it.\n");
              System.out.flush();
              mod.breakSolver();
            }
            return 0;
          }

      };
    }
  /*TAG:end-callback*/                
    public static void main(String[] args)
    {
      String slvr     = "intpnt";
      if (args.length < 1)
      {
        System.out.println("Usage: ( psim | dsim | intpnt ) ");
      }

      if (args.length >= 1) slvr     = args[0];

        double[][] A = 
            { { 3.0, 2.0, 0.0, 1.0 },
              { 2.0, 3.0, 1.0, 1.0 },
              { 0.0, 0.0, 3.0, 2.0 } };
        double[] c = { 3.0, 5.0, 1.0, 1.0  };
        
        double maxtime = 0.01;

        try(Model M = new Model("callback"))
        {
          Variable x = M.variable("x", 3, Domain.greaterThan(0.0));
          Variable y = M.variable("y", 1, Domain.inRange(0.0, 10.0));
          Variable z = Var.vstack(x,y);
          M.constraint("c1", Expr.dot(A[0], z), Domain.equalsTo(30.0));
          M.constraint("c2", Expr.dot(A[1], z), Domain.greaterThan(15.0));
          M.constraint("c3", Expr.dot(A[2], z), Domain.lessThan(25.0));

          M.objective("obj", ObjectiveSense.Maximize, Expr.dot(c, z));

          if( slvr == "psim") 
            M.setSolverParam("optimizer", "primal_simplex");
          else if( slvr == "dsim")
            M.setSolverParam("optimizer", "dual_simplex");
          else if( slvr == "intpnt")
            M.setSolverParam("optimizer", "intpnt");

          /*TAG:begin-callback-handler*/                
          M.setCallbackHandler( makeUserCallback(maxtime,M) );
          /*TAG:end-callback-handler*/                
          M.setSolverParam("log", 0);

          M.solve();

          }
      catch (mosek.Exception e)
      {
        System.out.println ("An error/warning was encountered");
        System.out.println (e.toString());
        throw e;
      }
   }
}

File: public/.julia/v0.5/mosek/deps/src/mosek/8/tools/examples/fusion/dotnet/callback.cs
//
//   Copyright: Copyright (c) MOSEK ApS, Denmark. All rights reserved.
//
//   File:      callback.cs
//
//   Purpose:   To demonstrate how to use the progress 
//              callback. 
//
//              Use this script as follows:
//              callback psim
//              callback dsim
//              callback intpnt
//
//              The first argument tells which optimizer to use
//              i.e. psim is primal simplex, dsim is dual simplex
//              and intpnt is interior-point. 
//

using System;
using mosek.fusion;

namespace mosek
{
  namespace fusion
  {
    namespace example
    {
      /*TAG:begin-callback*/                      
      class myCallback : mosek.DataCallback
      {
        double maxtime;
        Model M;

        public myCallback( double maxtime_,
                           Model M_)
          {
            maxtime = maxtime_;
            M = M_;
          }
        
        public override int callback( callbackcode caller, 
                                      double[]     douinf,
                                      int[]        intinf,
                                      long[]       lintinf )
          {
            double opttime = 0.0;
            int itrn;
            double pobj, dobj, stime;
  /*TAG:begin-callback-switch*/             
            switch (caller)
            {
            case callbackcode.begin_intpnt:
              Console.WriteLine("Starting interior-point optimizer");
              break;
            case callbackcode.intpnt:
              itrn    = intinf[(int) iinfitem.intpnt_iter      ];
              pobj    = douinf[(int) dinfitem.intpnt_primal_obj];
              dobj    = douinf[(int) dinfitem.intpnt_dual_obj  ];
              stime   = douinf[(int) dinfitem.intpnt_time      ];
              opttime = douinf[(int) dinfitem.optimizer_time   ];
            
              Console.WriteLine("Iterations: {0,-3}",itrn);
              Console.WriteLine("  Elapsed: Time: {0,6:F2}({1:F2})",opttime,stime);
              Console.WriteLine("  Primal obj.: {0,-18:E6}  Dual obj.: {1,018:E6}e",pobj,dobj);
              break;
            case callbackcode.end_intpnt:
              Console.WriteLine("Interior-point optimizer finished.");
              break;
            case callbackcode.begin_primal_simplex:
              Console.WriteLine("Primal simplex optimizer started.");
              break;
            case callbackcode.update_primal_simplex:
              itrn    = intinf[(int) iinfitem.sim_primal_iter  ];
              pobj    = douinf[(int) dinfitem.sim_obj          ];
              stime   = douinf[(int) dinfitem.sim_time         ];
              opttime = douinf[(int) dinfitem.optimizer_time   ];
            
              Console.WriteLine("Iterations: {0,-3}}", itrn);
              Console.WriteLine("  Elapsed time: {0,6:F2}({1:F2})",opttime,stime);
              Console.WriteLine("  Obj.: {0,-18:E6}", pobj );
              break;
            case callbackcode.end_primal_simplex:
              Console.WriteLine("Primal simplex optimizer finished.");
              break;
            case callbackcode.begin_dual_simplex:
              Console.WriteLine("Dual simplex optimizer started.");
              break;
            case callbackcode.update_dual_simplex:
              itrn    = intinf[(int) iinfitem.sim_dual_iter    ];
              pobj    = douinf[(int) dinfitem.sim_obj          ];
              stime   = douinf[(int) dinfitem.sim_time         ];
              opttime = douinf[(int) dinfitem.optimizer_time   ];
              Console.WriteLine("Iterations: {0,-3}}", itrn);
              Console.WriteLine("  Elapsed time: {0,6:F2}({1:F2})",opttime,stime);
              Console.WriteLine("  Obj.: {0,-18:E6}", pobj );
              break;
            case callbackcode.end_dual_simplex:
              Console.WriteLine("Dual simplex optimizer finished.");
              break;
            case callbackcode.begin_bi:
              Console.WriteLine("Basis identification started.");
              break;
            case  callbackcode.end_bi:
              Console.WriteLine("Basis identification finished.");
              break;
            default:
              break;
            }
  /*TAG:end-callback-switch*/ 
            if (opttime >= maxtime)
            {
              Console.WriteLine("MOSEK is spending too much time. Terminate it.");
              M.BreakSolver();
            }
            return 0;
          }
      }

  /*TAG:end-callback*/                      

      public class callback
      {
        public static void Main(string[] args)
          {
            
            string slvr     = "intpnt";
            
            if (args.Length < 1)
            {
              Console.WriteLine("Usage: callback ( psim | dsim | intpnt )");
            }

            if (args.Length >= 1) slvr = args[0];

            double[][] A =
              { new double[] { 3.0, 2.0, 0.0, 1.0 },
                new double[] { 2.0, 3.0, 1.0, 1.0 },
                new double[] { 0.0, 0.0, 3.0, 2.0 } };
            double[] c = { 3.0, 5.0, 1.0, 1.0  };

            double maxtime = 0.01;
            
            Model M = new Model("callback");
            
            Variable x = M.Variable("x", 3, Domain.GreaterThan(0.0));
            Variable y = M.Variable("y", 1, Domain.InRange(0.0, 10.0));
            Variable z = Var.Vstack(x,y);
            
            M.Constraint("c1", Expr.Dot(A[0], z), Domain.EqualsTo(30.0));
            M.Constraint("c2", Expr.Dot(A[1], z), Domain.GreaterThan(15.0));
            M.Constraint("c3", Expr.Dot(A[2], z), Domain.LessThan(25.0));
            
            M.Objective("obj", ObjectiveSense.Maximize, Expr.Dot(c, z));
            
                     
            if( slvr == "psim") 
              M.SetSolverParam("optimizer", "primal_simplex");
            else if( slvr == "dsim")
              M.SetSolverParam("optimizer", "dual_simplex");
            else if( slvr == "intpnt")
              M.SetSolverParam("optimizer", "intpnt");

            /*TAG:begin-callback-handler*/                            
            M.SetCallbackHandler( new myCallback(maxtime, M) );
            /*TAG:end-callback-handler*/                
            M.SetSolverParam("log", 0);
            
            M.Solve();
            
          }
      }
    }
  }
}
